scanner/updater/export.go

```diff
@@ -0,0 +1,180 @@
+package updater
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/quay/claircore/alpine"
+	"github.com/quay/claircore/aws"
+	"github.com/quay/claircore/debian"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/libvuln/jsonblob"
+	"github.com/quay/claircore/libvuln/updates"
+	"github.com/quay/claircore/oracle"
+	"github.com/quay/claircore/photon"
+	"github.com/quay/claircore/rhel"
+	"github.com/quay/claircore/suse"
+	"github.com/quay/claircore/ubuntu"
+	"github.com/quay/claircore/updater/osv"
+	"github.com/quay/zlog"
+	"github.com/stackrox/stackrox/scanner/v4/updater/manual"
+	"golang.org/x/time/rate"
+)
+
+// Export is responsible for triggering the updaters to download Common Vulnerabilities and Exposures (CVEs) data
+// and then outputting the result as a zstd-compressed file with .ztd extension
+func Export(ctx context.Context, outputDir string) error {
+	var updaters []driver.Updater
+
+	// Append updater sets directly to the updaters
+	appendUpdaterSet := func(updaterSet driver.UpdaterSet, err error) {
+		if err != nil {
+			zlog.Error(ctx).Err(err).Send()
+			return
+		}
+		updaters = append(updaters, updaterSet.Updaters()...)
+	}
+
+	appendUpdaterSet(aws.UpdaterSet(ctx))
+	appendUpdaterSet(manual.UpdaterSet(ctx, nil))
+	appendUpdaterSet(oracle.UpdaterSet(ctx))
+	appendUpdaterSet(photon.UpdaterSet(ctx))
+	appendUpdaterSet(suse.UpdaterSet(ctx))
+
+	alpineFac, err := alpine.NewFactory(ctx)
+	if err != nil {
+		return err
+	}
+
+	debianFac, err := debian.NewFactory(ctx)
+	if err != nil {
+		return err
+	}
+
+	rhelFac, err := rhel.NewFactory(ctx, rhel.DefaultManifest)
+	if err != nil {
+		return err
+	}
+
+	ubuntuFac, err := ubuntu.NewFactory(ctx)
+	if err != nil {
+		return err
+	}
+
+	osvFac := osv.Factory
+
+	cfgs := map[string]driver.ConfigUnmarshaler{
+		"debian": func(v interface{}) error {
+			v.(*debian.FactoryConfig).MirrorURL = `https://deb.debian.org/`
+			return nil
+		},
+		"alpine": func(v interface{}) error {
+			v.(*alpine.FactoryConfig).URL = "https://secdb.alpinelinux.org/"
+			return nil
+		},
+		"rhel": func(v interface{}) error {
+			return nil
+		},
+		"ubuntu": func(v interface{}) error {
+			v.(*ubuntu.FactoryConfig).Name = "ubuntu"
+			return nil
+		},
+		"osv": func(v interface{}) error {
+			cfg := v.(*osv.FactoryConfig)
+			cfg.URL = osv.DefaultURL
+			return nil
+		},
+	}
+
+	facs := map[string]driver.UpdaterSetFactory{
```

Comment:
Is the idea here just to have a subset of updaters run with their default configurations? If so I would take a look at how `clairctl` does it's export: https://github.com/quay/clair/blob/main/cmd/clairctl/export.go#L136 there seems to be a lot that isn't needed. Also, not really sure why the in-tree updaters are being passed as out-of-tree updaters. Can you expand on what memory issues you faced and how initialization looked when you experienced those issues?

Reply:
https://gist.github.com/crozzy/f51aecf9efd0e89b51dad2f2c27f9085 this might make things a little clearer

Reply:
```
dataStore, err := jsonblob.New()
	updateMgr, err := updates.NewManager(ctx, dataStore, updates.NewLocalLockSource(), httpClient,
updates.WithConfigs(cfgs),
		updates.WithFactories(facs),
		updates.WithOutOfTree(updaters),
	)
```
I encountered issues while attempting to trigger all updaters within a single update manager using one JSON blob. The workflow gets cancelled quite often. I am guessing even the output file is about 100 MB but the json blob can be consuming quite a few memory when invoking `err = dataStore.Store(zstdWriter)` ?